llvm/lib/Transforms/Vectorize/LoopVectorize.cpp

```diff
@@ -3994,8 +3996,13 @@ LoopVectorizationCostModel::computeMaxVF(ElementCount UserVF, unsigned UserIC) {
   }
 
   unsigned TC = PSE.getSE()->getSmallConstantTripCount(TheLoop);
-  unsigned MaxTC = PSE.getSE()->getSmallConstantMaxTripCount(TheLoop);
+
+  SmallVector<const SCEVPredicate *, 2> Predicates;
+  unsigned MaxTC =
+      PSE.getSE()->getSmallConstantMaxTripCount(TheLoop, &Predicates);
```

Comment:
Should this use PSE.getSmallConstantMaxTripCount() to ensure the predicates are added to PSE?

Reply:
Hmm, that's a good question. I'd assumed that when we call PSE.getBackedgeTakenCount(TheLoop, &Predicates) at some point later on to create the trip count, that it would just end up adding the same predicates anyway.

Does it matter if we attempt to add the same predicate twice? If not, then I'm happy to change this and use the predicated version.

Reply:
I think it's safe to add the predicates multiple times, since the `addPredicate` function in PredicatedScalarEvolution already checks to see if the new predicate is already implied by its existing list. I've added a new `getSmallConstantMaxTripCount` to PredicatedScalarEvolution and kept track of the count to avoid unnecessary computation, similar to how we return the backedge taken count.

Reply:
Yes, PSE should take care to avoid adding duplicated predicates, thanks for adjusting!

Reply:
Should a similar change also be done for the recent changes for early exit checks in LV?

Reply:
Do you mean the calls to `PSE.getSE()->getPredicatedExitCount` in `LoopVectorizationLegality::isVectorizableEarlyExitLoop`? I'm not sure if we want to do exactly the same, i.e. cache the exit count for every exit. I could potentially add an equivalent method to `PredicatedScalarEvolution` that automatically adds the predicates, but probably best done in a separate patch because it's unrelated to the max trip count. What do you think?